aidial_adapter_bedrock/llm/truncate_prompt.py

```diff
@@ -0,0 +1,106 @@
+from abc import ABC, abstractmethod
+from typing import Callable, List, Optional, Set, TypeVar
+
+from pydantic import BaseModel
+
+from aidial_adapter_bedrock.utils.list import select_by_indices
+
+
+class TruncatePromptError(ABC, BaseModel):
+    @abstractmethod
+    def print(self) -> str:
+        pass
+
+
+class InconsistentLimitsError(TruncatePromptError):
+    user_limit: int
+    model_limit: int
+
+    def print(self) -> str:
+        return (
+            f"Maximum prompt tokens ({self.user_limit}) "
+            f"exceeds the model maximum prompt tokens ({self.model_limit})."
+        )
+
+
+class ModelLimitOverflow(TruncatePromptError):
+    model_limit: int
+    token_count: int
+
+    def print(self) -> str:
+        return (
+            f"Token count of all messages ({self.token_count}) exceeds"
+            f" the model maximum prompt tokens ({self.model_limit})."
+        )
+
+
+class UserLimitOverflow(TruncatePromptError):
+    user_limit: int
+    token_count: int
+
+    def print(self) -> str:
+        return (
+            "Token count of the last message and all system messages "
+            f"({self.token_count}) exceeds the maximum prompt tokens ({self.user_limit})."
+        )
+
+
+T = TypeVar("T")
+Messages = List[T]
+
+
+def truncate_prompt(
+    messages: Messages,
+    tokenize: Callable[[Messages], int],
+    keep_message: Callable[[Messages, int], bool],
+    model_limit: Optional[int],
+    user_limit: Optional[int],
+) -> Set[int] | TruncatePromptError:
```

Comment:
Why not to throw an exception from the method instead?

Reply:
1. Because the error doesn't represent an exceptional situation, but it's a completely valid result of the truncation algorithm.

    If I throw an exception instead, then at every call site I need to make a decision:
    * Wrap the method call in try-catch or not?
    * If wrap, then which exceptions represent user errors and which - the internal errors?
   
    This makes the method a leaking abstraction leaning to a tightly coupled code.

    If I return user errors as part of the return value, then I don't need to think about any of this.

2. Returning an error as a value is much more convenient for the `truncate_prompt` endpoint, because it actually returns truncation errors as part of its response.